### ISSUE TITLE
NWX Continuous Integration

### ISSUE THREAD

**Issue by keipertk** [NWChemEx-Project/DeveloperTools#3]
I think that [Jenkins](https://jenkins.io/) is a better choice than TravisCI for NWX continuous integration. Here are some of the relative advantages and disadvantages of Jenkins:

Advantages:

- Jenkins servers can be deployed on any system with Java. With the wide range of target platforms for NWX, this gives us more freedom for automated testing on different hardware. (TravisCI cloud hardware is very slow/weak, by the way)

- You don't have to rebuild all dependencies every time you trigger a Jenkins build

- Since we control the Jenkins server, we can use any variety of software we choose (different versions of MPI, Intel compilers, etc.). TravisCI support of vendor tools seems to be limited, often due to license issues. See [here](https://github.com/travis-ci/travis-ci/issues/4604) for an example discussion.

- Like TravisCI, Jenkins can be fully integrated with Github (trigger builds on pull requests, report build status to Github, etc.)

- Like TravisCI, Jenkins supports Docker containers.

- Hundreds of plugins which support features such as static code analysis and code coverage reports. There are hooks for many external web tools which we've previously discussed. 

- We can run as many builds concurrently as we'd like.

- No job timeout limitations like with Travis

- Free, robust, open source project.

Disadvantages:

- Since Jenkins servers are deployed manually, we need reliable computer hardware for testing. At the moment I am running Jenkins on my laptop.

- There is more configuration work required compared to Travis, but it is pretty straightforward. I had everything up and running after a few hours of work.

I will illustrate some of the Jenkins capabilities now.

On my laptop, I execute a Jenkins .war file with java and specify HTTP port 8080. Then I open my browser and set it to localhost:8080. The web interface shows a single project which I've configured [here](https://i.imgur.com/1IHplXL.png).

The project I've created scrapes my Github account looking for any repository branch which contains a file in the root directory called "Jenkinsfile". (We could configure a project which checks every repo in the NWX organization.) The Jenkinsfile contains a script which tells Jenkins what to do when a build is triggered. To demonstrate, I added a Jenkinsfile to a my local fork of Ryan's UtilitiesEx repository - so only one repository is listed under my project [here](https://i.imgur.com/aju5Pu3.png). 

Clicking on the repository, I see a list of the branches. I've added the Jenkinsfile to only the Combinations branch. Clicking that, I see a graphical summary of the last few builds I ran [here](https://i.imgur.com/VOZHjt3.png). You can see that only my last build passed. The "Building", "Test", and "Deploy" stages were specified by me in the Jenkinsfile. We can choose any number of stages and specify any commands we want to be executed during our stages.

My Jenkinsfile is [here](https://i.imgur.com/yhxIjQ9.png). I execute the cmake build with simple shell commands. The test stage could contain a make check call, or something else to execute all Catch tests.

As far as monitoring build status/logs, this can all be integrated with Github. For example, we could just output the build error messages directly in the pull request. The build log can be accessed from the web server interface as well. You can find a portion of my successful build [here](https://i.imgur.com/rn9vOlJ.png).

In short, Jenkins provides more features and supports a broader hardware/software test matrix than TravisCI - at the cost of maintaining the test server ourselves. Thoughts?
